onnxruntime/core/optimizer/propagate_cast_ops.cc

```diff
@@ -1419,29 +1451,24 @@ Status PropagateCastOps::ApplyImpl(Graph& graph, bool& modified, int graph_level
     LOGS(logger, INFO) << "Propagate Cast operations summary:";
     LOGS(logger, INFO) << "Number of passes = " << pass;
     LOGS(logger, INFO) << "Nodes Inserted:";
-    std::for_each(inserted_node_names.begin(), inserted_node_names.end(), [removed_node_names, logger](std::string name) {
-      if (removed_node_names.find(name) == removed_node_names.end()) { LOGS(logger, INFO) << name; } });
+    std::for_each(inserted_nodes.begin(), inserted_nodes.end(), [&](NodeIndex idx) { LOGS(logger, INFO) << graph.GetNode(idx)->Name(); });
 
     LOGS(logger, INFO) << "Nodes Removed:";
-    std::for_each(removed_node_names.begin(), removed_node_names.end(), [logger](std::string name) {
-      if (inserted_node_names.find(name) == inserted_node_names.end()) { LOGS(logger, INFO) << name; } });
+    std::for_each(removed_node_names.begin(), removed_node_names.end(), [&](const std::string& name) { LOGS(logger, INFO) << name; });
 
     LOGS(logger, INFO) << "Nodes Converted to FP16:";
-    std::for_each(converted_node_names.begin(), converted_node_names.end(), [removed_node_names, logger](std::string name) {
-      if (removed_node_names.find(name) == removed_node_names.end() && inserted_node_names.find(name) == inserted_node_names.end()) {
-        LOGS(logger, INFO) << name; } });
+    std::for_each(converted_nodes.begin(), converted_nodes.end(), [&](NodeIndex idx) { LOGS(logger, INFO) << graph.GetNode(idx)->Name(); });
   }
-  inserted_node_names.clear();
-  converted_node_names.clear();
   return Status::OK();
 }
 
 PropagateCastOps::PropagateCastOps(GraphTransformerConfiguration::PropagateCastOpsConfiguration::Strategy strategy,
-                                   size_t level, const std::vector<std::string>& _allow_list,
+                                   size_t level, const std::vector<std::string>& allow_list,
                                    const std::unordered_set<std::string>& compatible_execution_providers) noexcept
-    : GraphTransformer("PropagateCastOps", compatible_execution_providers), level_(level), strategy_(strategy) {
-  fp16_allow_ops[0].clear();  // Remove previously added op types if any.
-  std::copy(_allow_list.begin(), _allow_list.end(), std::inserter(fp16_allow_ops[0], fp16_allow_ops[0].begin()));
+    : GraphTransformer("PropagateCastOps", compatible_execution_providers), level_(level), fp16_allow_ops_(3), strategy_(strategy) {
```

Comment:
nit: lines below are kind of long.
if this is the only place that level1_fp16_allow_ops/level2_fp16_allow_ops are used, how about just initializing them here?
```suggestion
    : GraphTransformer("PropagateCastOps", compatible_execution_providers), level_(level),
      fp16_allow_ops_{
          std::unordered_set<std::string>(allow_list.begin(), allow_list.end()),
          {"level1 op", ... },
          {"level2 op", ... }
      },
      strategy_(strategy) {
```
and then fp16_allow_ops_ can be const too

Reply:
I want to keep the level1 and level2 operators at the top of the file so that it is easy to modify (add new operators and delete existings ones.) and/or add a new levels of fp16 operator all together.

Reply:
if you want to keep them separate, it's fine too.
```c++
constexpr std::array level1_fp16_allow_ops = {"...", ... }; // deduced to std::array<const char*, N>, don't need to specify N
...
  // ctor member initializer list
  fp16_allow_ops_{
      std::unordered_set<std::string>(allow_list.begin(), allow_list.end()),
      std::unordered_set<std::string>(level1_fp16_allow_ops.begin(), level1_fp16_allow_ops.end()),
      std::unordered_set<std::string>(level2_fp16_allow_ops.begin(), level2_fp16_allow_ops.end())
  }
```
as is, the lines in the ctor are longer than 120 characters

Reply:
Making the suggested change resulted in the following error:
/home/sajandhy/onnxruntime/onnxruntime/core/optimizer/propagate_cast_ops.cc:1468:269:   required from here
/usr/include/c++/9/bits/hashtable.h:994:4: error: no matching function for call to ‘std::_Hashtable<std::__cxx11::basic_string<char>, std::__cxx11::basic_string<char>, std::allocator<std::__cxx11::basic_string<char> >, std::__detail::_Identity, std::equal_to<std::__cxx11::basic_string<char> >, std::hash<std::__cxx11::basic_string<char> >, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, true, true> >::insert(const std::basic_string_view<char>&)’